### ISSUE TITLE
Stabilize control

### ISSUE THREAD

**Issue by AndreWeiner** [FabianGabriel/Active_flow_control_past_cylinder_using_DRL#6]
Hi Fabian,

the next task will be to stabilize the learning process by constraining the angular velocity. There are two means to restrict the angular velocity:

1. limiting the angular velocity applied as the boundary condition
2. limiting the distribution from which omega is sampled during training

The first point is the most essential one since extreme values of omega ultimately make the simulations crash. Restricting the exploration/the distribution should lead to a more efficient sampling, and therefore, to a smoother and faster training.

To implement step 1., you need to make modifications in the *agentRotatingWallVelocityFvPatchVectorField.C* file and the *get_predictions* method of the policy network for computing a normalized omega. You find a reference implementation on page 394 of grokking DRL:
```
mean, log_std = self.forward(state)
pi_s = Normal(mean, log_std.exp())
pre_tanh_action = pi_s.rsample()
tanh_action = torch.tanh(pre_tanh_action)
action = self.rescale_fn(tanh_action)
log_prob = pi_s.log_prob(pre_tanh_action) - torch.log(
            (1 - tanh_action.pow(2)).clamp(0, 1) + epsilon)
log_prob = log_prob.sum(dim=1, keepdim=True)
``` 
Note that the reason for having to implement the same code both in Python and C++ is that, at the time of writing, tracing probability distributions is not supported ([read more](https://github.com/pytorch/pytorch/issues/29843)).

Tip: Before running a training, make sure that both implementations consistently deliver outputs in the prescribed range.

As always, if questions or problems come up, feel free to post them here.

Best, Andre

**Comment by FabianGabriel** [comment]
Hi Andre,

I have now implemented this code in both Python and C++, and while the outputs are in the prescriped range, the limiting of the log_std valueis somewhat questionable.
Starting at a standard deviation value of about 0.8, the distribution shifts more towards the edges, making the mean less likely than other values:
![tanh-lower](https://user-images.githubusercontent.com/64479822/130588796-0c68abae-f137-4b05-86ea-9a54130939e8.png)

At even higher values, the distribution shifts even more drastically to the edges, cresting a shape like this:
![tanh-high](https://user-images.githubusercontent.com/64479822/130589002-5cc1f0a3-62a4-4d89-b66f-7100058a9f58.png)

For our machine learning case here this behaviour would be not intended and even counterproductive, right?
So to counteract this, I would suggest changing the log_std_max value in the boundary condition to -0.35 (approx. ln(0.7)).

Best Regards
Fabian

**Comment by AndreWeiner** [comment]
Hi Fabian,
that's indeed an important observation. Could you please create a notebook/script with the following three plots of the limited Gaussian distribution?

- influence of std
- influence of mean
- combined influence of mean and std

Thanks!
Andre

**Comment by FabianGabriel** [comment]
Hi Andre,

I just uploaded a notebook here: [notebook](https://github.com/FabianGabriel/Active_flow_control_past_cylinder_using_DRL/blob/main/notebooks/tanh_distribution.ipynb)

Best Regards,
Fabian

**Comment by AndreWeiner** [comment]
Good job! Very interesting profiles but their potential impact on the training is not so easy to foresee (at least for me). Clipping too early might corrupt the gradient needed to update the log_std output of the policy network. Let's give the value you suggested a try and see if the learning still works.
Best, Andre

**Comment by FabianGabriel** [comment]
Hi Andre,

I still have some problems with calculating the entropy of the bounded distribution.
The implementation of the sampling, squashing with tanh and then rescaling part in the _agentRotatingWallVelocityFvPatchVectorField.C_ file was pretty straight forward and there is also a formula there for the log_probability. But I can't seem to find a way to calculate the entropy of this changed distrinution.

The implementation in the _get_predictions_ part was also a little bit more challeging as that function is getting actions and states as their input. So I can't just sample new actions there. But as thoose actions are read from an existing _trajectory.csv_ file, I can be sure that those actions are already bounded to the actionspace. To calculate the log_probability there I had to "descale" the actions and reverse the tanh function. Then I can use the formula from above too. That _should_ now also lead to the correct log probabilities.
But I also have to calculate the entropies there, so I have essentially the same problem there too.

Best Regards,
Fabian

**Comment by FabianGabriel** [comment]
Hi Andre,

on Friday we talked about the tanh-distribution and how the integral is not 1.
Well, turns out I made a typo in the plotting notebook. The integral actually _is_ 1.
I'm sorry and I uploaded the fixed version of the notebook.

Best Regards,
Fabian

**Comment by AndreWeiner** [comment]
Hi @FabianGabriel,

great that you fixed the issue! I've tried to derive the entropy equation for bounded normal distribution, but the outcome was too complex to be useful (there were a couple of terms without a closed-form solution). Maybe a more advanced tool like Mathematica can compute the integral. I've gone through the procedure once for the normal distribution, and I append the result here for completeness:
![PXL_20210830_101229626](https://user-images.githubusercontent.com/8482575/131340074-e34673cf-d11e-43c2-98ad-1e23fcbf09b6.jpg)
As I mentioned in the meeting, the exact shape of the entropy term is not crucial as far as I understand. If you plot the differential entropy of the normal distribution, you see that this term in the policy loss function basically prevents the standard deviation to drop to zero (the consequence of which would be no exploration).

Regarding the implementation of the beta distribution in the OpenFOAM boundary condition:
- [this source](http://www.cplusplus.com/reference/random/gamma_distribution/) shows how to use the gamma distribution in the c++ standard library
- [this source](https://stackoverflow.com/questions/10358064/random-numbers-from-beta-distribution-c) describes that one can sample Z ~ Beta(a,b) from X ~ Gamma(a,1) and Y ~ Gamma(b,1) as Z=X/(X+Y)

Here is a minimal working example (I scaled the output by ten to use the original plotting function):
```
// compile with
// g++ beta.C -o beta
#include <iostream>
#include <random>

int main()
{
  const int nrolls=10000;  // number of experiments
  const int nstars=100;    // maximum number of stars to distribute

  std::default_random_engine generator;
  std::gamma_distribution<double> distribution_1(2.0, 1.0);
  std::gamma_distribution<double> distribution_2(2.0, 1.0);

  int p[10]={};

  for (int i=0; i<nrolls; ++i) {
    double number_1 = distribution_1(generator);
    double number_2 = distribution_2(generator);
    double sample = number_1 / (number_1+number_2) * 10;
    ++p[int(sample)];
  }

  std::cout << "beta distribution (2.0,2.0):" << std::endl;

  for (int i=0; i<10; ++i) {
    std::cout << i << "-" << (i+1) << ": ";
    std::cout << std::string(p[i]*nstars/nrolls,'*') << std::endl;
  }

  return 0;
}
```
If you need further suggestions for the implementation in OpenFOAM, let me know.
Best, Andre

**Comment by FabianGabriel** [comment]
Hi Andre,

thank you for theese suggestions. I already thought the use of the 2 gamma distributions was the way to go here and implemented a very similar solution. (For the record: The second number in the 2 gamma distributions doesn't have to be 1 as long as they are the same: Z ~ Beta(a,b) from X ~ Gamma(a,c) and Y ~ Gamma(b,c), c can be any number here).

I have now implemented both the tanh-normal-distribution and the beta-distribution and started a training each. The results look very promising so far.

Best Regards,
Fabian
